Enable not emptyable arguments for is-empty and not-empty clauses

lbrdnk opened the pull request:
Closes #40885, #40883,  #41005

### Description

This PR enables `:is-empty` arguments to be just`::expression/expression` (ie. do not have to be `::expression/emptyable`). Then, check if the arg is emptyable is moved to desugar code. For not emptyable args, `:is-empty` is expanded the same way as `:is-null`.

This PR also renames `test-data-with-null-date-checkins` dataset to `test-data-null-date`. Reason being that recently we've added logic into `metabase.test.data.interface/db-qualified-table-name` that requires resulting names to be shorther than 30 chars. Because of that, dataset was not usable with eg. redshift, vertica, oracle... And now it is.

### How to verify

Follow the reproduction steps from linked issues or run updated tests.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
Enable not emptyable arguments for is-empty and not-empty clauses

bshepherdson commented on src/metabase/legacy_mbql/util.cljc:
```diff
@@ -159,12 +159,33 @@
+  [[_ _id-or-name {:keys [base-type]} :as field]]
+  (or base-type
+      ;; TODO: How to make following work in JS?
+      #?(:clj (when (integer? _id-or-name)
```
I think this is overkill. `:base-type` should be included on every field ref these days.

This smarter desugaring of `:is-empty` is a courtesy to users who use `:is-empty` when they meant `:is-null`.

In the case where we don't already have the `:base-type` (which I'm not sure can even happen outside tests that create bad field refs), let's just preserve the old behavior and return the string-y desugaring like `[:or [:= field nil] [:= field ""]]`. Then there's no need to look up fields in the metadata.

lbrdnk replied:
I did as suggested. I've also added guard in `add-type-info` string implementation, to handle un-probable case of field having no base type in options, hence being treated as `emptyable` and so empty string arg (in expansion) would be coerced to date/time type. Approach is further described in the new test.

lbrdnk committed:
Add review remarks

lbrdnk committed:
Shorten dataset name so it is usable with some drivers

lbrdnk committed:
Disable test for Athena

lbrdnk committed:
Handle possibly missing base type

lbrdnk commented on src/metabase/query_processor/middleware/wrap_value_literals.clj:
```diff
@@ -223,7 +223,8 @@
   (if (and unit
-           parse-datetime-strings?)
+           parse-datetime-strings?
+           (seq s))
```
This avoids attempt of coercion of empty string.

lbrdnk commented on test/metabase/test/data/dataset_definitions.clj:
```diff
@@ -112,8 +112,8 @@
 (defonce ^{:doc "The main `test-data` dataset, with an additional (all-null) `null_only_date` Field."}
-  test-data-with-null-date-checkins
-  (tx/transformed-dataset-definition "test-data-with-null-date-checkins" test-data
+  test-data-null-date
+  (tx/transformed-dataset-definition "test-data-null-date" test-data
```
Renaming of dataset is explained in PR description.

snoe commented on test/metabase/query_processor_test/filter_test.clj:
```diff
@@ -765,6 +765,54 @@
+(deftest ^:parallel is-empty-not-empty-with-not-emptyable-args-test
+  (mt/test-drivers
+   ;; TODO: Investigate how to make the test work with Athena!
+   (disj (mt/normal-drivers) :athena)
```
Take a look at docs in `metabase.test.data.athena`

snoe commented on src/metabase/lib/schema/filter.cljc:
```diff
@@ -75,7 +75,7 @@
 ;;; :not-empty is sugar for [:and [:!= ... nil] [:!= ... ""]]
```
update comment

snoe commented on src/metabase/lib/schema/filter.cljc:
```diff
@@ -75,7 +75,7 @@
 ;;; :not-empty is sugar for [:and [:!= ... nil] [:!= ... ""]]
 (doseq [op [:is-empty :not-empty]]
   (mbql-clause/define-tuple-mbql-clause op :- :type/Boolean
-    [:ref ::expression/emptyable]))
```
is emptyable still in use?

snoe commented on src/metabase/legacy_mbql/util.cljc:
```diff
@@ -159,12 +159,27 @@
+    [:not-empty field]
+    (if (emptyable-expansion? field)
+      [:and [:!= field nil] [:!= field ""]]
+      [:!=  field nil])))
```
nit: extra space

lbrdnk committed:
Update comment

lbrdnk committed:
Missing base type field is treated as not emptyable

lbrdnk committed:
Update desugar middleware

Now middleware adds base type to fields with integer id that are missing that first. This ensures correct function of is-empty expansion.

lbrdnk committed:
Update desugar middleware

Cleanup base types necessary for desugaring is-empty

lbrdnk committed:
Update desugar middleware cleanup

lbrdnk committed:
Update test

lbrdnk committed:
Adjust tests

lbrdnk commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -4,21 +4,54 @@
+   [metabase.query-processor.store :as qp.store]
    [metabase.util.malli :as mu]))
 
+(defn- add-missing-field-base-types
```
This function ensures that fields with id and missing base type have base in field options prior to dusugaring. That is necessary for correct `:is-empty` and `:not-empty` desugaring.

lbrdnk commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -4,21 +4,54 @@
-
-                                      (expression :guard mbql.preds/FieldOrExpressionDef?)
-                                      (mbql.u/desugar-expression expression)))))
+  (m/update-existing query :query (comp remove-desugar-added-base-types
+                                        desugar*
+                                        add-missing-field-base-types)))
```
Base types are added to field options for integer id fields that are missing those. Then original desugar is performed. Later those base types (added for desugaring purposes) are removed.

bshepherdson replied:
This comment should go in the code, not the review :)

lbrdnk commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -4,21 +4,54 @@
   `inside` with lower-level clauses like `between`. This is done to minimize the number of MBQL clauses individual
   drivers need to support. Clauses replaced by this middleware are marked `^:sugar` in the MBQL schema."
   [query]
-  (m/update-existing query :query (fn [query]
-                                    (lib.util.match/replace query
-                                      (filter-clause :guard mbql.preds/Filter?)
-                                      (mbql.u/desugar-filter-clause filter-clause)
-
-                                      (temporal-extract-clause :guard mbql.preds/DatetimeExpression?)
-                                      (mbql.u/desugar-temporal-extract temporal-extract-clause)
-
-                                      (expression :guard mbql.preds/FieldOrExpressionDef?)
-                                      (mbql.u/desugar-expression expression)))))
```
Desugar code was moved to standalone fn.

lbrdnk commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -4,21 +4,54 @@
+   [metabase.query-processor.store :as qp.store]
    [metabase.util.malli :as mu]))
 
+(defn- add-missing-field-base-types
+  "Add base type to fields with id that are missing it. It is necessary for correct function
+   of [[metabase.legacy-mbql.util/desugar-is-empty-and-not-empty]]."
+  [query]
+  (lib.util.match/replace query
+    (field-clause :guard (fn [clause]
+                           (and (mbql.preds/Field? clause)
+                                (integer? (second clause))
+                                (not (contains? (get clause 2) :base-type)))))
+    (let [id (second field-clause)
+          metadata (lib.metadata/field (qp.store/metadata-provider) id)
+          {:keys [base-type]} metadata]
+      (update field-clause 2 assoc
+              :base-type base-type
+              ::desugar-added-base-type true))))
```
This function ensures that fields with id and missing base type have base in field options prior to dusugaring. That is necessary for correct :is-empty and :not-empty desugaring.

bshepherdson commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -4,21 +4,54 @@
+(defn- remove-desugar-added-base-types
+  [query]
+  (lib.util.match/replace query
+    (field-clause :guard (fn [clause]
+                           (and (mbql.preds/Field? clause)
+                                (integer? (second clause))
+                                (get-in clause [2 ::desugar-added-base-type]))))
+    (update field-clause 2 (comp not-empty
+                                 #(dissoc % :base-type ::desugar-added-base-type)))))
```
I think this would be easier to read with some destructuring in the `:guard`, `(fn [[_field id opts :as clause]] ...)`.

lbrdnk replied:
Guard fn could be called on something that's not destructurable. That would result in an exception. Hence second and get-in must be called after the Field? check.

lbrdnk committed:
Remove now redundant e2e test

lbrdnk committed:
Update comment

lbrdnk committed:
Merge branch 'master' into escalation-coerce-temporal
